pkg/kubectl/cmd/util/clientcache.go

```diff
@@ -78,31 +78,46 @@ func (c *ClientCache) ClientConfigForVersion(version string) (*client.Config, er
 	}
 	config.GroupVersion = negotiatedVersion
 	client.SetKubernetesDefaults(&config)
-	c.configs[version] = &config
+
+	if version != nil {
+		c.configs[*version] = &config
+	}
 
 	// `version` does not necessarily equal `config.Version`.  However, we know that we call this method again with
 	// `config.Version`, we should get the the config we've just built.
 	configCopy := config
-	c.configs[config.GroupVersion.String()] = &configCopy
+	c.configs[*config.GroupVersion] = &configCopy
 
 	return &config, nil
 }
 
 // ClientForVersion initializes or reuses a client for the specified version, or returns an
 // error if that is not possible
-func (c *ClientCache) ClientForVersion(version string) (*client.Client, error) {
-	if client, ok := c.clients[version]; ok {
-		return client, nil
+func (c *ClientCache) ClientForVersion(version *unversioned.GroupVersion) (*client.Client, error) {
+	if version != nil {
+		if client, ok := c.clients[*version]; ok {
+			return client, nil
+		}
 	}
 	config, err := c.ClientConfigForVersion(version)
 	if err != nil {
 		return nil, err
 	}
-	client, err := client.New(config)
+
+	kubeclient, err := client.New(config)
 	if err != nil {
 		return nil, err
 	}
+	c.clients[*config.GroupVersion] = kubeclient
+
+	if version != nil {
```

Comment:
I see it's because next time `ClientConfigForVersion()` will return you with the same config if the `version` is the same. A comment here can be useful. Up to you.